### PR TITLE
soc: arm: atmel: Introduce SAMC21x

### PR DESCRIPTION
**Pull request by benys**
At https://github.com/zephyrproject-rtos/hal_atmel/pull/27 I pushed hal for SAM C21

Signed-off-by: Kamil Serwus <kserwus@gmail.com>

### CODE REVIEW

**Comment by benys** [comment]
Hi @martinjaeger it is no problem to split changes. I would like to clarify:
a) create soc/board without CAN & ADC
b) create standalone changes to ADC (to fix compliation) and add ADC definition
c) after merge a) & b) add CAN driver?

No problem for me, but I need wait for HAL approval - becouse many problems with compliation is caused by no HAL files.

**Comment by martinjaeger** [comment]
Yes, looks like a good approach. Whether the SOC and board stuff need to be split further (I've seen some new clock-control bindings, but no driver implementation) would be up to the ATSAM maintainer @nandojve.

Please note that we are currently in feature freeze for v3.2 release, so the new additions can only be merged after the release. Assuming the ADC fix is a bug, it could be merged before the release already.

**Comment by benys** [comment]
@martinjaeger ok I created #50388.
I am waiting for approve HAL and #50388 to make future PR (with only soc/board and CAN)

**Comment by henrikbrixandersen** [comment]
> a) create soc/board without CAN & ADC
> b) create standalone changes to ADC (to fix compliation) and add ADC definition
> c) after merge a) & b) add CAN driver?

Sounds good. The main reason for splitting into several PRs, as pointed out by others, is to keep the size of the PR down - but you also need to consider that different features like these are likely to be reviewed by different groups of people. By splitting the PR into logically separate changes you also allow for the PRs to have the reviewers assigned that really care for and know the details of the topic in question.

**Comment by benys** [comment]
@martinjaeger @henrikbrixandersen I removed from commit ADC and CAN releated changes. This is pure C21 support

File: boards/arm/atsamc21n_xpro/doc/index.rst
Comment (martinjaeger) [inline boards/arm/atsamc21n_xpro/doc/index.rst:15]: The file is named atsam**d** instead of **c**. Is this a typo?

**Comment by martinjaeger** [comment]
> @martinjaeger @henrikbrixandersen I removed from commit ADC and CAN releated changes. This is pure C21 support

Thanks! Handing over to @nandojve as I'm not familiar with ATSAM hardware. But I'll be happy to review the follow-up CAN related PR.

File: boards/arm/atsamc21n_xpro/doc/index.rst
Comment (benys) [reply boards/arm/atsamc21n_xpro/doc/index.rst:15]: @martinjaeger atmel created two xpro boards: "ATSAMC21N-XPRO" and "ATSAMC21-XPRO". I tested with ATSAMC21N-XPRO

https://www.microchip.com/en-us/products/microcontrollers-and-microprocessors/32-bit-mcus/sam-32-bit-mcus/sam-c

Comment (benys) [reply boards/arm/atsamc21n_xpro/doc/index.rst:15]: Yes I will fix typo :)

**Comment by nandojve** [comment]
Hi @benys ,

Thank you for your contribution. I`ll look on it as soon I can.

**Comment by benys** [comment]
> > @martinjaeger @henrikbrixandersen I removed from commit ADC and CAN releated changes. This is pure C21 support
> 
> Thanks! Handing over to @nandojve as I'm not familiar with ATSAM hardware. But I'll be happy to review the follow-up CAN related PR.

I created #50408 PR

**Comment by benys** [comment]
Hello @nandojve

Did you have some time to look to my PR :)?

Kamil

**Comment by benys** [comment]
I rebase to current main branch.

Anything I should change?

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro.yaml
Comment (nandojve) [inline boards/arm/atsamc21n_xpro/atsamc21n_xpro.yaml:20]: Are you sure that `counter`, `dma` and `watchdog` are running on this board?

File: boards/arm/atsamc21n_xpro/Kconfig.board
Comment (nandojve) [inline boards/arm/atsamc21n_xpro/Kconfig.board:1]: Please, split board into their own commit. It should be last on PR series.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @benys,

It is important make sure this board is running on CI tests. You can run locally `./scripts/twister --platform atsamc21n_xpro`

**Comment by benys** [comment]
> Hi @benys,
> 
> It is important make sure this board is running on CI tests. You can run locally `./scripts/twister --platform atsamc21n_xpro`

It looks that it pass

File: boards/arm/atsamc21n_xpro/doc/index.rst
Comment (benys) [reply boards/arm/atsamc21n_xpro/doc/index.rst:116]: I read similar board. My intention was to point out that SERCOM5 is configured as SPI. (The board description point that)
But you are right that I made mistake.

I remove this section - becouse board doesn't have external flash.

File: west.yml
Comment (benys) [reply west.yml:60]: I created new PR: https://github.com/zephyrproject-rtos/zephyr/pull/51226

File: boards/arm/atsamc21n_xpro/doc/index.rst
Comment (nandojve) [reply boards/arm/atsamc21n_xpro/doc/index.rst:43]: You should add an entry like below to keep it. But you must test it to make sure driver is OK.
https://github.com/zephyrproject-rtos/zephyr/blob/c27a6af7128a433fca825c43a0efe2f9405892b5/boards/arm/atsamd20_xpro/atsamd20_xpro.dts#L95-L111

File: boards/arm/atsamc21n_xpro/doc/index.rst
Comment (nandojve) [inline boards/arm/atsamc21n_xpro/doc/index.rst:43]: Please, update table to new model, like SAMR34.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @benys,

You need rebase. There are things that need to be addressed here before we merge hal. We need make sure everything is ok on the hal before merge it.

**Comment by benys** [comment]
> Hi @benys,
> 
> You need rebase. There are things that need to be addressed here before we merge hal. We need make sure everything is ok on the hal before merge it.

Ok, no problem. I will prepare all things:)

I thought that i can wait for merging Hal.

File: soc/arm/atmel_sam0/common/Kconfig.samc2x
Comment (benys) [reply soc/arm/atmel_sam0/common/Kconfig.samc2x:6]: yes, it is place for configuration - for example clocks etc.

File: west.yml
Comment (benys) [reply west.yml:60]: I will create separated commit in this PR

**Comment by benys** [comment]
@nandojve I think that I have last problem with this PR. HAL defines C20, C20N, C21, C21N folders (series).

So I discovered that this implicates that I need make C2x and C2x soc series.

For example:
```
config SOC_SERIES
	default "samc21n" if SOC_PART_NUMBER_SAMC21N17A
	default "samc21n" if SOC_PART_NUMBER_SAMC21N18A
	default "samc21"
```
so this implicates configuration SOC_SERIES_SAMC21 and SOC_SERIES_SAMC21N.
Is it ok?

**Comment by nandojve** [comment]
> @nandojve I think that I have last problem with this PR. HAL defines C20, C20N, C21, C21N folders (series).
> 
> So I discovered that this implicates that I need make C2x and C2x soc series.
> 
> ```
> /zephyrproject/zephyr/soc/arm/atmel_sam0/samc21/soc.h:48:10: fatal error: samc21n18a.h: No such file or directory
>    48 | #include <samc21n18a.h>
>       |          ^~~~~~~~~~~~~~
> ```
> 
> becouse atmel hal searches includes like this:
> 
> ```
> zephyr_include_directories(${CONFIG_SOC_SERIES})
> ```
> 
> So I tried
> 
> ```
> config SOC_SERIES
> 	default "samc21n" if SOC_PART_NUMBER_SAMC21N17A
> 	default "samc21n" if SOC_PART_NUMBER_SAMC21N18A
> 	default "samc21"
> ```
> 
> so this implicates configuration SOC_SERIES_SAMC21 and SOC_SERIES_SAMC21N. Even I discovered that maybe I need create separated samc21 samc21n folder in soc/atmel/atmel_sam0
> 
> But easier way maybe is change at atmel hal include for something like
> 
> ```
> if c21xxn or c20xxn
> zephyr_include_directories(${CONFIG_SOC_SERIES}n)
> else 
> zephyr_include_directories(${CONFIG_SOC_SERIES})
> ```

Yes, there is this implication. However, we need add a solution that may solve future problems without give us more work.
Please add a commit (after the update manifest) with following content and I as co-author:

```diff
diff --git a/soc/arm/atmel_sam0/Kconfig b/soc/arm/atmel_sam0/Kconfig
index ebda1a5179..a11cf9b95a 100644
--- a/soc/arm/atmel_sam0/Kconfig
+++ b/soc/arm/atmel_sam0/Kconfig
@@ -18,5 +18,6 @@ source "soc/arm/atmel_sam0/common/Kconfig.saml2x"
 source "soc/arm/atmel_sam0/common/Kconfig.samd2x"
 source "soc/arm/atmel_sam0/common/Kconfig.samd5x"
 source "soc/arm/atmel_sam0/*/Kconfig.soc"
+source "soc/arm/atmel_sam0/Kconfig.soc.revisions"
 
 endif
diff --git a/soc/arm/atmel_sam0/Kconfig.soc.revisions b/soc/arm/atmel_sam0/Kconfig.soc.revisions
new file mode 100644
index 0000000000..a152528ac7
--- /dev/null
+++ b/soc/arm/atmel_sam0/Kconfig.soc.revisions
@@ -0,0 +1,10 @@
+# Copyright (c) 2022 Gerson Fernando Budke <nandojve@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_REVISION_N
+       bool
+
+config SOC_SERIES_REVISION
+       string
+       default "n" if SOC_SERIES_REVISION_N
+       default ""
```

Add this patch at hal_atmel:

```diff
diff --git a/asf/sam0/include/CMakeLists.txt b/asf/sam0/include/CMakeLists.txt
index d273adb..77e86df 100644
--- a/asf/sam0/include/CMakeLists.txt
+++ b/asf/sam0/include/CMakeLists.txt
@@ -1 +1 @@
-zephyr_include_directories(${CONFIG_SOC_SERIES})
+zephyr_include_directories(${CONFIG_SOC_SERIES}${CONFIG_SOC_SERIES_REVISION})
```

Then, on your commits you can just add:

```diff
diff --git a/soc/arm/atmel_sam0/samc21/Kconfig.soc b/soc/arm/atmel_sam0/samc21/Kconfig.soc
index 3f82572539..f3f324abf6 100644
--- a/soc/arm/atmel_sam0/samc21/Kconfig.soc
+++ b/soc/arm/atmel_sam0/samc21/Kconfig.soc
@@ -51,8 +51,10 @@ config SOC_PART_NUMBER_SAMC21J18AU
 
 config SOC_PART_NUMBER_SAMC21N17A
        bool "SAMC21N17A"
+       select SOC_SERIES_REVISION_N
 
 config SOC_PART_NUMBER_SAMC21N18A
        bool "SAMC21N18A"
+       select SOC_SERIES_REVISION_N
 
 endchoice
```

**Comment by benys** [comment]
@nandojve nice solution. I am still learning zephyr philosophy :)

File: dts/arm/atmel/samc20.dtsi
Comment (nandojve) [inline dts/arm/atmel/samc20.dtsi]: There is no need to keep it if it is empty.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @benys,

Overall LGTM, a few styles are missing and I would like confirm that flash storage is OK.
When CI is green I will merge hal_atmel. After that, you need replace PR number for the hash.

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts
Comment (benys) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:140]: I will give you feedback/output and I will check this sample

Comment (benys) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:140]: @nandojve sample works if I declare in sample prj.conf - CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES=y

So I should do anything or add this config to board?

After added:
```
*** Booting Zephyr OS build zephyr-v3.2.0-885-g05f643cc7a42  ***
No address found, adding 192.168.1.1 at id 1
No key found, adding it at id 2
No Reboot counter found, adding it at id 3
Id: 4 not found, adding it
Longarray not found, adding it as id 5
[00:00:00.005,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 2
[00:00:00.005,000] <inf> fs_nvs: No GC Done marker found: restarting gc
[00:00:00.005,000] <dbg> fs_nvs: nvs_flash_erase_sector: Erasing flash at 3c200, len 256
[00:00:00.007,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 0
[00:00:00.007,000] <dbg> fs_nvs: nvs_add_gc_done_ate: Adding gc done ate at f0
[00:00:00.009,000] <dbg> fs_nvs: nvs_flash_erase_sector: Erasing flash at 3c000, len 256
[00:00:00.010,000] <inf> fs_nvs: 3 Sectors of 256 bytes
[00:00:00.010,000] <inf> fs_nvs: alloc wra: 2, e8
[00:00:00.010,000] <inf> fs_nvs: data wra: 2, 0
[00:00:00.010,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.015,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.019,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.022,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.026,000] Reboot counter history: ...0
Oldest reboot counter: 0
Rebooting in ...5<dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.030,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.034,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.036,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.040,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.044,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
[00:00:00.152,000] <dbg> fs_nvs: nvs_recover_last_ate: Recovering last ate from sector 1
...4...3...2...1
```

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro-pinctrl.dtsi
Comment (benys) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro-pinctrl.dtsi:19]: @nandojve in this project how many spaces == tab?
In my vscode is aligned :)

File: soc/arm/atmel_sam0/common/Kconfig.samc2x
Comment (benys) [reply soc/arm/atmel_sam0/common/Kconfig.samc2x:6]: I will investigate - maybe I will add one/some configuration

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro-pinctrl.dtsi
Comment (benys) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro-pinctrl.dtsi:19]: i believe that 8 - i will fix

File: soc/arm/atmel_sam0/common/Kconfig.samc2x
Comment (benys) [reply soc/arm/atmel_sam0/common/Kconfig.samc2x:6]: I removed clock for RTC. I don't have scenerio for it.
Also PRO board doesn't have fast external oscilator so I don't want to implement and test it "by declare in code".

Maybe in feature I will propose (or somebode else) to add configuration :-)

**Comment by benys** [comment]
Hi @nandojve I think you can merge it ;)

After merge I want to extend my PR with CAN support (#50408) with this SOC:)

Thanks a lot!

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts
Comment (nandojve) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:140]:
```diff
diff --git a/drivers/flash/Kconfig.sam0 b/drivers/flash/Kconfig.sam0
index e604b0ce8e..accaa30d30 100644
--- a/drivers/flash/Kconfig.sam0
+++ b/drivers/flash/Kconfig.sam0
@@ -16,6 +16,8 @@ menuconfig SOC_FLASH_SAM0
 config SOC_FLASH_SAM0_EMULATE_BYTE_PAGES
        bool "Emulate byte-sized pages"
        depends on SOC_FLASH_SAM0
+       default y if SOC_SERIES_SAMC20 || \
+                    SOC_SERIES_SAMC21
        help
          Emulate a device with byte-sized pages by doing a
          read/modify/erase/write.
```

Comment (benys) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:140]: Yes I can add, but I believe that is the same problem with D2x mcus.
So you suggest that I should add this defaults releated with SAMC2x?

Comment (nandojve) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:140]: If you know this is necessary for any other SAM0, feel free to add the series. I mean, my understanding is that C2x only works with SOC_FLASH_SAM0_EMULATE_BYTE_PAGES=y so you need add it.

**Comment by nandojve** [comment]
> I see that PR wit ADC is completed. Maybe I will add devicetrees related with ADC in this PR in my third commit :)?

Yes, please do it. Make sure that tests are running using twister as I already mentioned for CI side and run a test on board to make sure everything is in place.

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts
Comment (benys) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:150]: I don't have other boards. :( So i live only c2x

File: boards/arm/atsamc21n_xpro/doc/index.rst
Comment (nandojve) [inline boards/arm/atsamc21n_xpro/doc/index.rst:46]: ADC is missing

File: boards/arm/atsamc21n_xpro/doc/index.rst
Comment (nandojve) [inline boards/arm/atsamc21n_xpro/doc/index.rst:89]: ADC is missing

**Review by nandojve — CHANGES_REQUESTED** [review]
Almost there...  : -)

**Comment by benys** [comment]
> Almost there... : -)

No problem, I am patient :-)

**Review by nandojve — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
